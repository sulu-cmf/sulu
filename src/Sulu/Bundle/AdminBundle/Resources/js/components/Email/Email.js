// @flow
import React from 'react';
import Isemail from 'isemail';
import {observer} from 'mobx-react';
import {action, computed, observable} from 'mobx';
import Input from '../Input';

type Props = {|
    name?: string,
    placeholder?: string,
    valid: boolean,
    value: ?string,
    onBlur?: () => void,
    onChange: (value: ?string) => void,
|};

@observer
export default class Email extends React.Component<Props> {
    static defaultProps = {
        valid: true,
    };

    @observable value: ?string;
    @observable showError: boolean = false;

    @action setValue(value: ?string) {
        this.value = value;
    }

    @action setShowError(showError: boolean) {
        this.showError = showError;
    }

    @computed get isValidValue(): boolean {
        if (!this.value) {
            return true;
        }

        return Isemail.validate(this.value);
    }

    componentDidMount() {
        this.setValue(this.props.value);
    }

    componentDidUpdate() {
        if (this.value && !this.props.value) {
            return;
        }

        this.setValue(this.props.value);
    }

    handleIconClick = () => {
        const {value} = this.props;
        if (!value) {
            return;
        }

        window.location.assign('mailto:' + value);
    };

    handleBlur = () => {
        if (this.isValidValue) {
            this.setShowError(false);
        } else {
            this.props.onChange(undefined);
            this.setShowError(true);
        }

        const {onBlur} = this.props;

        if (onBlur) {
            onBlur();
        }
    };

    handleChange = (value: ?string) => {
        this.setValue(value);

        if (!this.isValidValue) {
            this.props.onChange(undefined);

            return;
        }

        this.setShowError(false);
        this.props.onChange(this.value);
    };

    render() {
        const {
            valid,
            name,
            placeholder,
            value,
        } = this.props;

        return (
            <Input
                icon="su-envalope"
                onChange={this.handleChange}
                value={this.value}
                type="email"
                valid={valid && !this.showError}
                name={name}
                placeholder={placeholder}
                onBlur={this.handleBlur}
                onIconClick={(value && value.length > 1) ? this.handleIconClick : undefined}
            />
        );
    }
}
