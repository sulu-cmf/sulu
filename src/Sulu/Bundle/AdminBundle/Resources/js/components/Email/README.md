The email component can be used to get a valid email from the user.

```javascript
initialState = {value: undefined};
const onChange = (newValue) => {
    setState({value: newValue});
};

<div>
    <div style={{paddingBottom: '50px'}}>Current value: {state.value ? state.value : 'null'}</div>
    <Email value={state.value} onChange={onChange} />
</div>
```