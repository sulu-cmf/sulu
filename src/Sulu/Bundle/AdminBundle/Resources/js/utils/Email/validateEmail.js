// @flow

export default function(value: ?string): boolean {
    if (!value) {
        return false;
    }

    // used by webkit and recommend by the WHATWG:
    //  - https://html.spec.whatwg.org/#valid-e-mail-address
    //  - https://github.com/WebKit/WebKit/blob/WebKit-7615.1.12.130.1/Source/WebCore/html/EmailInputType.cpp#L38-L39

    // eslint-disable-next-line max-len
    return /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/.test(value);
}
