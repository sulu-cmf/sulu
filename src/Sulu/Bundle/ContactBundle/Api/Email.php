<?php

/*
 * This file is part of Sulu.
 *
 * (c) Sulu GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\ContactBundle\Api;

use JMS\Serializer\Annotation\Groups;
use JMS\Serializer\Annotation\SerializedName;
use JMS\Serializer\Annotation\VirtualProperty;
use Sulu\Bundle\ContactBundle\Entity\Email as EmailEntity;
use Sulu\Bundle\ContactBundle\Entity\EmailType as EmailTypeEntity;
use Sulu\Component\Rest\ApiWrapper;

class Email extends ApiWrapper
{
    public function __construct(EmailEntity $email, $locale)
    {
        $this->entity = $email;
        $this->locale = $locale;
    }

    /**
     * @VirtualProperty
     * @SerializedName("id")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getId(): ?int
    {
        return $this->entity->getId();
    }

    public function setEmail(string $email): self
    {
        $this->entity->setEmail($email);

        return $this;
    }

    /**
     * @VirtualProperty
     * @SerializedName("email")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getEmail(): ?string
    {
        return $this->entity->getEmail();
    }

    public function setEmailType(EmailTypeEntity $emailType): self
    {
        $this->entity->setEmailType($emailType);

        return $this;
    }

    /**
     * @VirtualProperty
     * @SerializedName("emailType")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getEmailType(): ?int
    {
        return $this->entity->getEmailType()->getId();
    }
}
