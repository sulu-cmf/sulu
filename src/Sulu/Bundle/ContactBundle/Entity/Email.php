<?php
/*
 * This file is part of the Sulu CMS.
 *
 * (c) MASSIVE ART WebServices GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\ContactBundle\Entity;
use JMS\Serializer\Annotation\Exclude;
use JMS\Serializer\Annotation\Groups;

/**
 * Email
 */
class Email
{
    /**
     * @var string
     * @Groups({"fullAccount", "partialAccount", "fullContact", "partialContact"})
     */
    private $email;

    /**
     * @var integer
     * @Groups({"fullAccount", "partialAccount", "fullContact", "partialContact"})
     */
    private $id;

    /**
     * @var \Sulu\Bundle\ContactBundle\Entity\EmailType
     * @Groups({"fullAccount", "fullContact"})
     */
    private $emailType;

    /**
     * @var \Doctrine\Common\Collections\Collection
     * @Exclude
     */
    private $contacts;

    /**
     * @var \Doctrine\Common\Collections\Collection
     * @Exclude
     */
    private $accounts;

    /**
     * Constructor
     */
    public function __construct()
    {
        $this->contacts = new \Doctrine\Common\Collections\ArrayCollection();
        $this->accounts = new \Doctrine\Common\Collections\ArrayCollection();
    }

    /**
     * Set email
     *
     * @param string $email
     * @return Email
     */
    public function setEmail($email)
    {
        $this->email = $email;

        return $this;
    }

    /**
     * Get email
     *
     * @return string
     */
    public function getEmail()
    {
        return $this->email;
    }

    /**
     * Get id
     *
     * @return integer
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Set emailType
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\EmailType $emailType
     * @return Email
     */
    public function setEmailType(\Sulu\Bundle\ContactBundle\Entity\EmailType $emailType)
    {
        $this->emailType = $emailType;

        return $this;
    }

    /**
     * Get emailType
     *
     * @return \Sulu\Bundle\ContactBundle\Entity\EmailType
     */
    public function getEmailType()
    {
        return $this->emailType;
    }

    /**
     * Add contacts
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\Contact $contacts
     * @return Email
     */
    public function addContact(\Sulu\Bundle\ContactBundle\Entity\Contact $contacts)
    {
        $this->contacts[] = $contacts;

        return $this;
    }

    /**
     * Remove contacts
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\Contact $contacts
     */
    public function removeContact(\Sulu\Bundle\ContactBundle\Entity\Contact $contacts)
    {
        $this->contacts->removeElement($contacts);
    }

    /**
     * Get contacts
     *
     * @return \Doctrine\Common\Collections\Collection
     */
    public function getContacts()
    {
        return $this->contacts;
    }

    /**
     * Add accounts
     *
     * @param AccountInterface $account
     * @return Email
     */
    public function addAccount(AccountInterface $account)
    {
        $this->accounts[] = $account;

        return $this;
    }

    /**
     * Remove accounts
     *
     * @param AccountInterface $account
     */
    public function removeAccount(AccountInterface $account)
    {
        $this->accounts->removeElement($account);
    }

    /**
     * Get accounts
     *
     * @return \Doctrine\Common\Collections\Collection
     */
    public function getAccounts()
    {
        return $this->accounts;
    }
}
