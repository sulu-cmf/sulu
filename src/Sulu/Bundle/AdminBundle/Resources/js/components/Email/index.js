// @flow
import Email from './Email';

export default Email;
