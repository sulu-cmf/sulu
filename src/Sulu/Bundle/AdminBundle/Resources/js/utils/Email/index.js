// @flow
import validateEmail from './validateEmail';

export {validateEmail};
