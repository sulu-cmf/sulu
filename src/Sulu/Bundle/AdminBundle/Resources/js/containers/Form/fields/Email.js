// @flow
import React from 'react';
import {default as EmailComponent} from '../../../components/Email';
import type {FieldTypeProps} from '../../../types';

export default class Email extends React.Component<FieldTypeProps<?string>> {
    render() {
        const {error, onChange, onFinish, value} = this.props;

        return (
            <EmailComponent
                onChange={onChange}
                onBlur={onFinish}
                valid={!error}
                value={value}
            />
        );
    }
}
