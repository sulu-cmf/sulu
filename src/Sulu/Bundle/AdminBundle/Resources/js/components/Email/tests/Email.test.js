// @flow
import React from 'react';
import {render, mount} from 'enzyme';
import Email from '../Email';

test('Email should render', () => {
    const onChange = jest.fn();
    expect(render(<Email value={null} onChange={onChange} />)).toMatchSnapshot();
});

test('Email should render with placeholder', () => {
    const onChange = jest.fn();
    expect(render(<Email value={null} placeholder="My placeholder" onChange={onChange} />)).toMatchSnapshot();
});

test('Email should render with value', () => {
    const onChange = jest.fn();
    const value = 'test@test.com';
    expect(render(<Email value={value} onChange={onChange} />)).toMatchSnapshot();
});

test('Email should render null value as empty string', () => {
    const onChange = jest.fn();
    expect(render(<Email value={null} onChange={onChange} />)).toMatchSnapshot();
});

test('Email should render error', () => {
    const onChange = jest.fn();
    expect(render(<Email value={null} onChange={onChange} valid={false} />)).toMatchSnapshot();
});

test('Email should render error when invalid value is set', () => {
    const onChange = jest.fn();
    const email = mount(<Email value={null} onChange={onChange} />);

    // check if showError is set correctly
    email.find('Input').instance().props.onChange('xxx', {target: {value: 'xxx'}});
    email.find('Input').instance().props.onBlur();
    email.update();
    expect(email.instance().showError).toBe(true);

    expect(render(email)).toMatchSnapshot();

    // now add a valid value
    email.find('Input').instance().props.onChange('test@test.com', {target: {value: 'test@test.com'}});
    email.find('Input').instance().props.onBlur();
    email.update();
    expect(email.instance().showError).toBe(false);

    expect(render(email)).toMatchSnapshot();
});

test('Email should trigger callbacks correctly', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const email = mount(<Email value={null} onChange={onChange} onBlur={onBlur} />);

    // provide invalid value
    email.find('Input').instance().props.onChange('xxx', {target: {value: 'xxx'}});
    email.find('Input').instance().props.onBlur();
    email.update();
    expect(onChange).toBeCalledWith(undefined);
    expect(onBlur).toBeCalled();

    // provide one more invalid value
    email.find('Input').instance().props.onChange('abc', {target: {value: 'abc'}});
    email.find('Input').instance().props.onBlur();
    email.update();
    expect(onChange).toBeCalledWith(undefined);
    expect(onBlur).toBeCalled();

    // now add a valid value
    email.find('Input').instance().props.onChange('test@test.com', {target: {value: 'test@test.com'}});
    email.find('Input').instance().props.onBlur();
    email.update();
    expect(onChange).toBeCalledWith('test@test.com');
    expect(onBlur).toBeCalled();

    expect(onBlur).toHaveBeenCalledTimes(3);
});

test('Email should not set onIconClick when value is invalid', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const email = mount(<Email value={null} onChange={onChange} onBlur={onBlur} />);

    expect(email.find('Input').prop('onIconClick')).toBeUndefined();
});

test('Email should set onIconClick when value is valid and window should be opened', () => {
    window.location.assign = jest.fn();

    const onChange = jest.fn();
    const onBlur = jest.fn();
    const email = mount(<Email value={'abc@abc.abc'} onChange={onChange} onBlur={onBlur} />);

    const onIconClickFunction = email.find('Input').prop('onIconClick');
    expect(onIconClickFunction).toBeInstanceOf(Function);
    onIconClickFunction.call();
    expect(window.location.assign).toBeCalledWith('mailto:abc@abc.abc');
});
