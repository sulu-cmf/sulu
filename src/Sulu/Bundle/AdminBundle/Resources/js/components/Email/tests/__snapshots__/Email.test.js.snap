// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email should render 1`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render error 1`] = `
<label
  class="input error"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render error when invalid value is set 1`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render error when invalid value is set 2`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render null value as empty string 1`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render with placeholder 1`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon su-envalope"
    />
  </div>
  <input
    placeholder="My placeholder"
    type="email"
    value=""
  />
</label>
`;

exports[`Email should render with value 1`] = `
<label
  class="input"
>
  <div
    class="prependedContainer"
  >
    <span
      aria-label="su-envalope"
      class="icon iconClickable su-envalope clickable"
      role="button"
      tabindex="0"
    />
  </div>
  <input
    type="email"
    value=""
  />
</label>
`;
